ding/framework/message_queue/torch_rpc.py

```diff
@@ -0,0 +1,365 @@
+
+            for timeout_peer in timeout_list:
+                self.peer_set.remove(timeout_peer)
+
+            return future_list
+
+    def publish_by_func(self, func: callable, *args, **kwargs) -> List[torch.Future]:
+        """
+        Overview:
+            Provide the upper-layer application with an interface for passing in
+            custom RPC methods, which is convenient for testing.
+        Arguments:
+            - func (callable): [Custom RPC methods]
+
+        Returns:
+            List: [The future objects list returned by the rpc_async call]
+        """
+        if self._running:
+            timeout_list = []
+            future_list = []
+
+            if len(self.peer_set) == 0:
+                logging.warning("No peer available to communicate with")
+                return []
+
+            # fix me: RuntimeError: Set changed size during iteration
+            for peer in self.peer_set:
+                if not self._running:
+                    break
+                try:
+                    if self._async:
+                        future_list.append(rpc.rpc_async(peer, func, args=(args), kwargs=kwargs))
+                    else:
+                        future_list.append(rpc.rpc_sync(peer, func, args=(args), kwargs=kwargs))
+                except RuntimeError as e:
+                    logging.error("Publish func to peer has error: \"{}\"!".format(peer))
+                    timeout_list.append(peer)
+
+            for timeout_peer in timeout_list:
+                self.peer_set.pop(timeout_peer)
+
+            return future_list
+
+    def accept_rpc_connect(self, peer_name: str, peer_rank: int, *args, **kwargs) -> bool:
+        """
+        Overview:
+            Receive the link signal sent by the peer.
+        Arguments:
+            - peer_name (str)
+            - peer_rank (int):
+
+        Returns:
+            bool: [Response status flag]
+        """
+        if peer_name in self.peer_set:
+            logging.warning(
+                "RPC server \"{}\" complain that received multiple stub requests from same client \"{}\"".format(
+                    self.name, peer_name
+                )
+            )
+        else:
+            self.peer_set.add(peer_name)
+
+        return True
+
+    def notify_shutdown(self, peer_name: str, *args, **kwargs) -> None:
+        """
+        Overview:
+            Receive the exit signal sent by the peer.
+        Arguments:
+            - peer_name (str)
+        """
+        if peer_name not in self.peer_set:
+            logging.warning(
+                "RPC server \"{}\" complain that received multiple stub requests from same client \"{}\"".format(
+                    self._mq.name, peer_name
+                )
+            )
+        else:
+            logging.warning("Recv shutdown info from \"{}\"".format(peer_name))
+            self.peer_set.remove(peer_name)
+
+    def subscribe(self, topic: str) -> None:
+        return
```

Comment:
raise NotImplementedError now

Reply:
Modified, now torhcrpc implements `subscribe()` method, so that the upper-level code can bypass the eventloop of `Task()` and `Parallel()`.